[MXNET-120] Float16 support for distributed training

rahul003 opened the pull request:
## Description ##
Supports distributed training with float16 gradients. 

- Modified parameter server to work at byte level, so that we can send both fp16 and fp32 gradients while training in fp16 and mixed precision modes. The user does not have to do anything special to handle fp16 gradients. KVStore handles it internally. Even originally, ps-lite later converted data to char, doing this at an earlier step gives mxnet more flexibility.

- The type of request and data type received by server are encoded in a command which is computed using the [Cantor pairing function](https://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function). This generates a unique number `z` given two numbers `x` and `y`, and allows us to invert this number `z` to give the unique original pair of numbers `x` and `y`.

- Server stores weights as fp32, because aggregating gradients and updating weights is about 20x slower on CPU if done as fp16. The cost of casting is much less than computing in fp16.

- To clarify, for keys with fp16 gradients, they are sent as fp16 and then cast to fp32 at server. Aggregating gradients and updating weights are done in fp32 on the server. When worker pulls weights for that key, the server would then return fp16 casted weights

- Fp16 training on 8 p3.8x machines for resent and vgg is now about *60% faster* than fp32. This seems consistent across different networks. 

## Checklist ##
### Essentials ###
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 

### Changes ###
- [x] Change kvdist and kvdist_server to receive float, half and double datatypes
- [x] Some sparse server side handling refactored into functions
- [x] Added nightly tests, and cleaned up some existing tests

### Comments ###
- Gradient compression is only supported for fp32. I'll fix them in a different PR. The compression code needs to be changed to handle different datatypes.

@eric-haibin-lin @reminisce @cjolivier01 @anirudh2290 @haojin2 @ptrendx please review

rahul003 commented:
@solin319 Please review this PR. This allows us to send some keys in fp16 and some in fp32. By using parameter server with type char, we also can avoid interface changes. Now the user doesn't have to do anything special to use fp16.

rahul003 commented:
@marcoabreu I see something weird in the CI build for this. 
I made some changes to address things raised by lint. But that broke compilation locally, yet the CI went through. Any idea what's going on?

marcoabreu commented:
In how far did it break compilation locally? what's your build environment?

haojin2 commented on src/kvstore/kvstore_dist.h:
```diff
@@ -619,15 +629,17 @@ class KVStoreDist : public KVStoreLocal {
+        pull_pskv.size *= num_bytes;
+        CHECK_EQ(static_cast<size_t>(push_pskv.size), compr_size * num_bytes);
+        CHECK_EQ(static_cast<size_t>(pull_pskv.size), original_size * num_bytes);
+        CHECK_EQ(push_pskv.lens.size(), num_servers * 2);
```
Can we not do something like:
CHECK_EQ(static_cast<size_t>(push_pskv.size), compr_size);
CHECK_EQ(static_cast<size_t>(pull_pskv.size), original_size);
push_pskv.size *= num_bytes;
pull_pskv.size *= num_bytes;
instead so that we perform 2 less multiplications?

rahul003 replied:
Good idea, will do

marcoabreu commented:
Huh? How come you closed the PR?

rahul003 commented:
@marcoabreu I had noticed a slowdown, which I have now fixed and reopened the PR

rahul003 commented:
I've changed my approach and added results section. Reviewers if you had already seen the code, please re-review it :)

rahul003 commented:
@mli Please also review the change to char parameter server we discussed. And for reference, the results section. Thanks!

marcoabreu commented on src/kvstore/kvstore_dist.h:
```diff
@@ -329,18 +330,21 @@ class KVStoreDist : public KVStoreLocal {
+          PSKV& pskv = EncodeDefaultKey(key, comm_buf.shape().Size(), true, num_bytes);
           PushDefault(key, comm_buf, pskv, priority);
         } else {
+          CHECK_EQ(dtype, mshadow::kFloat32) << "Gradient compression is only supported for "
```
Isn't it inconsistent to say Float16 is supported while actually not all features support it?

rahul003 replied:
This is the only feature left. It needs changes to some core gradient compression methods, which is coming in a later PR. Besides all changes *in kvstore* for gradient compression to support fp16 are there already.

marcoabreu replied:
Ah, sounds reasonable. Thanks!

rahul003 commented:
@marcoabreu I realized now that the reason CI passed then is because USE_DIST_KVSTORE is not on.

cjolivier01 commented:
That's going to be turned on, right?

rahul003 commented:
Yeah we should turn it on for all builds. @marcoabreu  Would submitting a regular PR (from a non-committer) use that script for PR build?

cjolivier01 commented on src/kvstore/kvstore_dist.h:
```diff
@@ -364,24 +368,23 @@ class KVStoreDist : public KVStoreLocal {
-      res_buf = NDArray(TShape{(int64_t) original_size}, comm_buf.ctx(),
-                        false, comm_buf.dtype());
+      small_buf = NDArray(TShape{pskv.size}, comm_buf.ctx(), false, dtype);
+      res_buf = NDArray(TShape{(int64_t) original_size}, comm_buf.ctx(), false, dtype);
```
nit: static_cast<>

cjolivier01 commented on src/kvstore/kvstore_dist.h:
```diff
@@ -636,7 +648,7 @@ class KVStoreDist : public KVStoreLocal {
   inline PSKV& EncodeRowSparseKey(const int key, const int64_t size, const int64_t num_rows,
                                   const int64_t *offsets, const size_t unit_len,
-                                  const int64_t total_num_rows) {
+                                  const int64_t total_num_rows, int num_bytes) {
```
nit: `const int num_bytes` (everything else is -- make it look like this has a special meaning since it isn't)

cjolivier01 replied:
throughout, yes pls try to reduce the superfluous computations

cjolivier01 commented on src/kvstore/kvstore_dist_server.h:
```diff
@@ -171,28 +210,31 @@ class KVStoreDistServer {
-            CHECK(updater_);
-            updater_(key, merged->array, stored);
-          });
+          CHECK(updater_);
```
nit: CHECK_NOTNULL() ?

rahul003 replied:
curious, is there a difference between `if (x)` vs `if ( x != NULL)`

rahul003 commented:
Thanks @cjolivier01 and @haojin2 , reduced some multiplications, added const to more places and changed couple of variable names ( num_bytes to elem_size and size to num_elems)

rahul003 commented:
@piiswrong Could you please review?

eric-haibin-lin commented on src/kvstore/kvstore_dist.h:
```diff
@@ -228,17 +228,18 @@ class KVStoreDist : public KVStoreLocal {
         // convert to ps keys
         size_t size = recv_buf.shape().Size();
-
+        int dtype = recv_buf.dtype();
```
const for `dtype`, `elem_size`, etc

eric-haibin-lin commented on src/kvstore/kvstore_dist.h:
```diff
@@ -228,17 +228,18 @@ class KVStoreDist : public KVStoreLocal {
         size_t size = recv_buf.shape().Size();
-
+        int dtype = recv_buf.dtype();
+        int elem_size = mshadow::mshadow_sizeof(dtype);
```
would `num_bytes` be a better name?

rahul003 replied:
Haha, that's what I had originally. I changed it from `num_bytes` to `elem_size` today evening! I changed it because I thought num_bytes was conveying the meaning of total number of bytes as opposed to the number of bytes for each element. Do you still think `num_bytes` is better?

eric-haibin-lin commented on src/kvstore/kvstore_dist.h:
```diff
@@ -506,41 +514,46 @@ class KVStoreDist : public KVStoreLocal {
    * \brief convert to keys in ps
    */
-  inline PSKV& EncodeDefaultKey(int key, size_t size, bool is_push) {
+  inline PSKV& EncodeDefaultKey(const int key, const size_t num_arr_elems,
```
Do you mind completing the param documentation for this function?

eric-haibin-lin commented on tests/nightly/dist_sync_kvstore.py:
```diff
@@ -278,39 +316,49 @@ def check_compr_random(kv, threshold, nworker):
-    test_sync_init()
-    test_sync_push_pull()
+    parser = argparse.ArgumentParser(description='test distributed kvstore in dist_sync mode')
+    parser.add_argument('--nrepeat', type=int, default=5)
```
Is the command in CI also updated?

marcoabreu replied:
@mbaijal manages nightly CI

rahul003 replied:
Same command as before can be used because the default values of the arguments runs all the earlier tests

marcoabreu commented:
@rahul003 Can't find the question, just have it in my email. I'll answer here:

> Yeah we should turn it on for all builds. @marcoabreu Would submitting a regular PR (from a non-committer) use that script for PR build?

Yes, every modification in a PR is picked up for that specific PR and will be applied to all PRs as soon as it has been merged.

rahul003 commented:
For reference: I'm adding the USE_DIST_KVSTORE flag to CI here

https://github.com/apache/incubator-mxnet/pull/10232

mli commented:
what's the default behavior for type converting? for example

- send fp16 grad to servers
- convert to fp16 grad to fp32 and then merge
- update fp32 weight with merged fp32 grad
- convert fp32 weight to fp16 and send to workers

rahul003 commented:
@mli
The behavior is as follows
If the grads pushed to kvstore are of type `dtype`
- Send grads in `dtype` form to server
- Server handles the gradients and maintains weights as 'dtype`. So the accumulation of grads as well as updating weights are in `dtype`
- Send weight back as `dtype`

This means if some keys are sent as fp16, they will be handled as fp16, and the server's copy of weights will be in fp16. If some keys are sent as fp32, they will be handled as fp32 and the server's copy of weights will be in fp32.

I experimented with storing weights in fp32 for all keys and casting received gradients into fp32, but the casting slowed things down greatly. My guess was the cast(copy) interferes with cache lines.

rahul003 commented:
@mli So if we cast to fp32, then we cast all received updates, as well as cast the weight after updating the weight so that we can respond to pulls in fp16. That was my original approach but it was slower. 

It looks like CopyFromTo already uses multiple threads, isn't it?

rahul003 commented:
Added multi-precision mode support. When optimizer's multi-precision field is True, then server maintains weights in fp32 and casts received gradients to fp32. 
Multi precision mode is slightly slower end-to-end[1]. In the experiments I ran it was about 10% slower than not using multi precision. But since it might be preferable to use multiprecision for convergence, I added it. I think there being both options is good because that makes it similar to single machine training.

[1] Using multi precision is about 10% slower than not using only when you compare end to end. When you profile server and also compare times for push and pull, the operations are almost 2x slower with multiprecision. CopyFromTo takes up a lot of time. I'll try updating CopyFromTo to use Kernel Launch instead of mshadow Expr and see if that makes a difference. But that is a different PR.

rahul003 commented on python/mxnet/kvstore.py:
```diff
@@ -474,6 +474,8 @@ def set_optimizer(self, optimizer):
                 raise
             self._send_command_to_servers(0, optim_str)
+            if optimizer.multi_precision:
+                self._send_command_to_servers(1, '')
```
changed

rahul003 commented on src/kvstore/kvstore_dist_server.h:
```diff
@@ -170,43 +216,90 @@ class KVStoreDistServer {
+
+        auto& update = update_buf_[key];
+        if (!update.merged.is_none()) {
+          update.merged = NDArray(update.merged.shape(), update.merged.ctx(), false,
```
you're right, added

rahul003 commented on src/kvstore/kvstore_dist_server.h:
```diff
@@ -170,43 +216,90 @@ class KVStoreDistServer {
+
+  inline void ApplyUpdates(const DataHandleType type, const int key,
+                           UpdateBuf *updateBuf, ps::KVServer<char>* server) {
+    if (!sync_mode_ || updateBuf->request.size() == (size_t) ps::NumWorkers()) {
```
done

rahul003 commented on src/kvstore/kvstore_dist_server.h:
```diff
@@ -220,175 +313,229 @@ class KVStoreDistServer {
+      return;
+    }
+    const NDArray& stored = store_[master_key];
+    if (has_multi_precision_copy(type)) stored.WaitToRead();
```
Because if there's a multi precision copy server returns after push when the multiprecision copy is ready. But while returning response to pull, we need the copy of stored in original datatype. I didn't add a wait in ApplyUpdates for this Copy because it's unnecessary at that point to wait. But before responding we need to wait for the copy to finish

eric-haibin-lin commented:
@marcoabreu are the nightly tests moved to apache CI already?

rahul003 commented:
An update, so the USE_DIST_KVSTORE flag has been added to the build, and this PR passes those builds on CPU and GPU.

marcoabreu commented:
@eric-haibin-lin no, not yet. @mbaijal is currently in the migration stage.

marcoabreu commented:
That's great, the only problem here is that CI is not running nightly tests as part of the PR validation. Could somebody remind me why we have to run KVStore tests in nightly instead of PR stage?

rahul003 commented:
So that means we can't merge it in for this release? This is an important feature which users are eager to use. We have requests from multiple teams who want to use this. I've validated that the nightly tests pass.

rahul003 commented:
Thanks, Do you know why the integration test doesn't show up in the CI steps?

marcoabreu commented:
The CI steps are cached and only re-evaluated if the stage is actually reached. It will show up once the job reaches the integration stage.

rahul003 commented:
I added the nightly tests we had for distributed kvstore as integration tests to the CI. And the build has passed. 

@everyone Please review and let me know if there are any more concerns

marcoabreu commented:
Thanks a lot, Rahul!

anirudh2290 commented:
@rahul003 can you fix the conflicts ?

rahul003 commented:
@eric-haibin-lin @piiswrong Is this good to be merged?

eric-haibin-lin commented on amalgamation/Makefile:
```diff
@@ -23,6 +23,11 @@ ifndef OPENBLAS_ROOT
 endif
 
+# use F16C if the architecture supports it, turned off by default
+ifndef USE_F16C
```
what does 16C stand for?

rahul003 replied:
added better comment. F16C is an instruction set extension supported by newer x86 CPUs. It provides intrinsics for faster fp16 compute.

eric-haibin-lin commented:
BTW build is failing

rahul003 commented:
Amalgamation build failed after I updated mshadow. I've updated the makefile for amalgamation to correct the build.

anirudh2290 commented:
@eric-haibin-lin can we merge this ?

anirudh2290 commented:
@piiswrong can this be merged ?
